Source Adapter Exception API (JavaScript)
===
Exception types in JavaScript source adapters are mapped to their ruby equivalent.  See the [ruby exception section](/rhoconnect/source-adapters#handling-exceptions) for more details about these exception types.

## Setup
Make sure you have a Rhoconnect Helpers object in your model:

    :::javascript
    var rc = require('rhoconnect_helpers');

## Exception(message)
Raise a generic RhoConnect adapter exception.

    :::javascript
    throw new rc.Exception('Something went wrong!');

## LoginException(message)
Raise an exception on the login procedure.

    :::javascript
    throw new rc.LoginException('A login error occured.');

## LogoffException(message)
Raise an exception on the logoff procedure.

    :::javascript
    throw new rc.LogoffException('A logoff error occured.');

## ServerTimeoutException(message)
Raise an exception indicating a timeout occured.

    :::javascript
    throw new rc.ServerTimeoutException('A timeout occured.');

## ServerErrorException(message)
Raise an exception concerning the backend service.

    :::javascript
    throw new rc.ServerErrorException('An error occured with our systems, please try again.');

## ObjectConflictErrorException(message)
Raise a [conflict exception](/rhoconnect/cud-conflicts) to indicate there was a conflict in handling the record. This will trigger RhoConnect's automatic processing for conflicts.

    :::javascript
    throw new rc.ObjectConflictErrorException(
      'A conflict error occured, please update your copy and try again'
    );


## Sample
The following example raises a `ServerErrorException` in the [delete function](/rhoconnectapi/source-adapter-model-api-js#del) if the record was not successfully deleted.

    :::javascript
    this.del = function(resp){
      var objId = resp.params.delete_object.id;
      var options = {
        host: host,
        path: '/products/' + objId + '.json',
        method: 'DELETE',
        headers: { 'Content-Type': 'application/json' }
      };
      var req = http.request(options, function(res){
        res.on('data', function(){});
        res.on('end', function(){
          resp.send(true);
        });
        res.on('error', function(){
          throw new rc.ServerErrorException('There was an error deleting, please try again');
        });
      });
      req.end();
    };
